### PR TITLE
MC-23881: When category URL page param is larger than the pagination, the page will break with elastic search 2

### LINKED ISSUE #23843: When category URL page param is larger than the pagination, the page will break with elastic search 2
### Preconditions (*)
1. Magento 2.3.2
1. Elasticsearch 2

### Steps to reproduce (*)
1. Create a category with products
1. Enable elasticsearch 2
1. Reindex all
1. Go to the category and add this params: "?p=999999999999999&product*list_order=price". Ex: https://www.example.com/category?p=9999999999999999999999999999&product_list*order=price

### Expected result (*)
1. Go to the last page of the category

### Actual result (*)
1. {{1 exception(s):
Exception #0 (Elasticsearch\Common\Exceptions\ServerErrorResponseException): {"error":{"root*cause":<{"type":"json_parse_exception","reason":"Numeric value (2.3058430092137e<ins>20) out of range of int (-2147483648 - 2147483647)\n at [Source: [B@33145eaa; line: 1, column: 28]"}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[{"shard":0,"index":"magento2_product_1_v173","node":"viwpe1vCTrGQomOe0uHakA","reason":{"type":"json_parse*exception","reason":"Numeric value (2.3058430092137e</ins>20) out of range of int (-2147483648 - 2147483647)\n at [Source: [B@33145eaa; line: 1, column: 28]"}}>},"status":500}}}

### PR DESCRIPTION
**Pull request by engcom-Golf**
### Fixed Issues (if relevant)

1. Fixes magento/magento2#23843

### Manual testing scenarios (*)

Create a category with products
Enable elasticsearch
Reindex all
Go to the category and add this param: ?p=999999999999999
Expected result:
No out of range exception in logs with status 500

Note:
There still can be log entity with 400 status.
Redirect to the last page of the category is not expected.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Golf** [comment]
@magento run all tests

File: app/code/Magento/Elasticsearch/Elasticsearch5/SearchAdapter/Query/Builder.php [hunk 1]
```diff
@@ -83,18 +85,18 @@ public function initQuery(RequestInterface $request)
     {
         $dimension = current($request->getDimensions());
         $storeId = $this->scopeResolver->getScope($dimension->getValue())->getId();
-
         $searchQuery = [
             'index' => $this->searchIndexNameResolver->getIndexName($storeId, $request->getIndex()),
             'type' => $this->clientConfig->getEntityType(),
             'body' => [
-                'from' => $request->getFrom(),
+                'from' => $this->getFrom($request),
```
Comment (ihor-sviziev) [inline app/code/Magento/Elasticsearch/Elasticsearch5/SearchAdapter/Query/Builder.php]: I think inline looks a bit cleaner. What do you thing?
```suggestion
                'from' => min((int)$request->getFrom(), self::ELASTIC_MAX_INT),
```

PS What about negative values? Should we also handle it?

File: app/code/Magento/Elasticsearch/SearchAdapter/Query/Builder.php [hunk 2]
```diff
@@ -61,7 +63,7 @@ public function initQuery(RequestInterface $request)
             'index' => $this->searchIndexNameResolver->getIndexName($storeId, $request->getIndex()),
             'type' => $this->clientConfig->getEntityType(),
             'body' => [
-                'from' => $request->getFrom(),
+                'from' => $this->getFrom($request),
```
Comment (ihor-sviziev) [inline app/code/Magento/Elasticsearch/SearchAdapter/Query/Builder.php]: same here

File: app/code/Magento/Elasticsearch/SearchAdapter/Query/Builder.php [hunk 3]
```diff
@@ -23,6 +23,8 @@
  */
 class Builder extends Elasticsearch5Builder
 {
+    private const ELASTIC_INT_MAX = 2147483647;
```
Comment (ihor-sviziev) [inline app/code/Magento/Elasticsearch/SearchAdapter/Query/Builder.php:26]:
```suggestion
    private const ELASTIC_MAX_INT = 2147483647;
```

File: app/code/Magento/Elasticsearch/Elasticsearch5/SearchAdapter/Query/Builder.php [hunk 4]
```diff
@@ -22,6 +22,8 @@
  */
 class Builder
 {
+    private const ELASTIC_INT_MAX = 2147483647;
```
Comment (ihor-sviziev) [inline app/code/Magento/Elasticsearch/Elasticsearch5/SearchAdapter/Query/Builder.php:25]:
```suggestion
    private const ELASTIC_MAX_INT = 2147483647;
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
one more question - on elastic 5+ it works fine?

File: app/code/Magento/Elasticsearch/Elasticsearch5/SearchAdapter/Query/Builder.php (hunk 4 above)
Comment (engcom-Golf) [reply app/code/Magento/Elasticsearch/Elasticsearch5/SearchAdapter/Query/Builder.php:25]: Agree that max_int sounds better but there is defined PHP_INT_MAX constant. I think it is better to keep the same format here.

File: app/code/Magento/Elasticsearch/Elasticsearch5/SearchAdapter/Query/Builder.php (hunk 1 above)
Comment (engcom-Golf) [reply app/code/Magento/Elasticsearch/Elasticsearch5/SearchAdapter/Query/Builder.php]: Agree. Will update. But don't think that we need to handle negative values. In that case, elastic returns exception with message numHits must be > 0

**Comment by engcom-Golf** [comment]
@ihor-sviziev fix tested for elastcsearch 6 and 7 versions

**Comment by engcom-Golf** [comment]
@magento run all tests

File: app/code/Magento/Elasticsearch/Elasticsearch5/SearchAdapter/Query/Builder.php (hunk 4 above)
Comment (ihor-sviziev) [reply app/code/Magento/Elasticsearch/Elasticsearch5/SearchAdapter/Query/Builder.php:25]: Ok, that make sense

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8618](https://jira.corp.magento.com/browse/ENGCOM-8618) has been created to process this Pull Request

**Review by zakdma — APPROVED** [review]

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
Was checked case from manual testing scenario and checked quick search and advanced search work as before
**Before:**
:x: Exception error in logs:
`[2021-01-12 08:33:52] main.CRITICAL: {"error":{"root_cause":[{"type":"input_coercion_exception","reason":"Numeric value (11999999999999976) out of range of int (-2147483648 - 2147483647)
 at [Source: (org.elasticsearch.common.io.stream.InputStreamStreamInput); line: 1, column: 26]"}],"type":"input_coercion_exception","reason":"Numeric value (11999999999999976) out of range of int (-2147483648 - 2147483647)
 at [Source: (org.elasticsearch.common.io.stream.InputStreamStreamInput); line: 1, column: 26]"},"status":500} {"exception":"[object] (Elasticsearch\\Common\\Exceptions\\ServerErrorResponseException(code: 500): {\"error\":{\"root_cause\":[{\"type\":\"input_coercion_exception\",\"reason\":\"Numeric value (11999999999999976) out of range of int (-2147483648 - 2147483647)\
 at [Source: (org.elasticsearch.common.io.stream.InputStreamStreamInput); line: 1, column: 26]\"}],\"type\":\"input_coercion_exception\",\"reason\":\"Numeric value (11999999999999976) out of range of int (-2147483648 - 2147483647)\
 at [Source: (org.elasticsearch.common.io.stream.InputStreamStreamInput); line: 1, column: 26]\"},\"status\":500} at /var/www/html/magento2dev/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:665)"} []
`

**After:**
:heavy_check_mark: Error with status code 400 that is acceptable
`[2021-01-12 09:10:00] main.CRITICAL: {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Result window is too large, from + size must be less than or equal to: [10000] but was [2147483659]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting."}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[{"shard":0,"index":"magento2dev_product_1_v109","node":"YDjpqEllRKOGLQ-KYNHWcQ","reason":{"type":"illegal_argument_exception","reason":"Result window is too large, from + size must be less than or equal to: [10000] but was [2147483659]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting."}}],"caused_by":{"type":"illegal_argument_exception","reason":"Result window is too large, from + size must be less than or equal to: [10000] but was [2147483659]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting.","caused_by":{"type":"illegal_argument_exception","reason":"Result window is too large, from + size must be less than or equal to: [10000] but was [2147483659]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting."}}},"status":400} {"exception":"[object] (Elasticsearch\\Common\\Exceptions\\BadRequest400Exception(code: 400): {\"error\":{\"root_cause\":[{\"type\":\"illegal_argument_exception\",\"reason\":\"Result window is too large, from + size must be less than or equal to: [10000] but was [2147483659]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting.\"}],\"type\":\"search_phase_execution_exception\",\"reason\":\"all shards failed\",\"phase\":\"query\",\"grouped\":true,\"failed_shards\":[{\"shard\":0,\"index\":\"magento2dev_product_1_v109\",\"node\":\"YDjpqEllRKOGLQ-KYNHWcQ\",\"reason\":{\"type\":\"illegal_argument_exception\",\"reason\":\"Result window is too large, from + size must be less than or equal to: [10000] but was [2147483659]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting.\"}}],\"caused_by\":{\"type\":\"illegal_argument_exception\",\"reason\":\"Result window is too large, from + size must be less than or equal to: [10000] but was [2147483659]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting.\",\"caused_by\":{\"type\":\"illegal_argument_exception\",\"reason\":\"Result window is too large, from + size must be less than or equal to: [10000] but was [2147483659]. See the scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level setting.\"}}},\"status\":400} at /var/www/html/magento2dev/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php:632)"} []
`
image

**Comment by engcom-Delta** [comment]
**Note:** Automation tests are passed

**Comment by zakdma** [comment]
@magento import code to https://github.com/magento-tsg/magento2ce

**Comment by magento-engcom-team** [comment]
@zakdma the branch with code successfully imported into`magento-tsg/magento2ce` repository. Branch name: `imported-magento-magento2-31594`.

**Pull request merged**